src/LifecycleManager.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace Vinelab\Bowler;
+
+use Closure;
+use PhpAmqpLib\Message\AMQPMessage;
+use Vinelab\Bowler\Exceptions\UnrecalledAMQPMessageException;
+
+class LifecycleManager
+{
+    /**
+     * @var array
+     */
+    protected $beforePublish = [];
+
+    /**
+     * @var array
+     */
+    protected $published = [];
+
+    /**
+     * @var array
+     */
+    protected $beforeConsume = [];
+
+    /**
+     * @var array
+     */
+    protected $consumed = [];
+
+    /**
+     * @param  Closure  $callback
+     * @return void
+     */
+    public function beforePublish(Closure $callback)
+    {
+        $this->beforePublish[] = $callback;
+    }
+
+    /**
+     * @param  Closure  $callback
+     * @return void
+     */
+    public function published(Closure $callback)
+    {
+        $this->published[] = $callback;
+    }
+
+    /**
+     * @param  Closure  $callback
+     * @return void
+     */
+    public function beforeConsume(Closure $callback)
+    {
+        $this->beforeConsume[] = $callback;
+    }
+
+    /**
+     * @param  Closure  $callback
+     * @return void
+     */
+    public function consumed(Closure $callback)
+    {
+        $this->consumed[] = $callback;
+    }
+
+    /**
+     * @param  AMQPMessage  $msg
+     * @param  string  $exchangeName
+     * @param  null  $routingKey
+     * @return AMQPMessage
+     * @throws UnrecalledAMQPMessageException
+     */
+    public function triggerBeforePublish(AMQPMessage $msg, string $exchangeName, $routingKey = null): AMQPMessage
+    {
+        foreach ($this->beforePublish as $callback) {
+            $msg = $callback($msg, $exchangeName, $routingKey);
+
+            if (!$msg instanceof AMQPMessage) {
+                throw new UnrecalledAMQPMessageException('Callback must return instance of AMQPMessage');
+            }
+        }
+
+        return $msg;
+    }
+
+    /**
+     * @param  AMQPMessage  $msg
+     * @param  string  $exchangeName
+     * @param  null  $routingKey
+     * @return void
+     */
+    public function triggerPublished(AMQPMessage $msg, string $exchangeName, $routingKey = null)
+    {
+        foreach ($this->published as $callback) {
+            $callback($msg, $exchangeName, $routingKey);
```

Comment:
should we be failsafe here? (e.g. `$callback` not having the correct number of arguments etc.)

Reply:
Executed callback are now [wrapped](https://github.com/Vinelab/bowler/pull/49/files#diff-eb2ac16609a9bd82b169a141bf74c71cR156) in a try-catch to ensure the execution proceeds in case of an error.